File: python/src/problems1_100/validate_binary_search_tree.py
from abc import ABC, abstractmethod
from typing import override

from src.concepts.tree_node import TreeNode


class Problem(ABC):
    @abstractmethod
    def is_valid_bst(self, root: TreeNode | None) -> bool:
        pass


class Default(Problem):
    @override
    def is_valid_bst(self, root: TreeNode | None) -> bool:
        return self.recursive_valid(root, None, None)

    def recursive_valid(self, node: TreeNode | None, left: int | None, right: int | None) -> bool:
        if not node:
            return True
        if left and node.val <= left:
            return False
        if right and node.val >= right:
            return False
        return self.recursive_valid(node.left, left, node.val) and \
            self.recursive_valid(node.right, node.val, right)


SOLUTIONS: list[Problem] = [Default()]

File: typescript/src/problems1_100/validate-binary-search-tree.ts
import TreeNode from '../concepts/tree-node';

class Problem {
  isValidBst(root: TreeNode | undefined): boolean {
    throw new Error(`Not implemented (${root}).`);
  }
}

class Default extends Problem {
  isValidBst(root: TreeNode | undefined): boolean {
    return this.recursiveValid(root, undefined, undefined);
  }

  recursiveValid(
      node: TreeNode | undefined,
      left: number | undefined,
      right: number | undefined,
  ): boolean {
    if (!node) {
      return true;
    }
    if (left && node.val <= left) {
      return false;
    }
    if (right && node.val >= right) {
      return false;
    }
    return this.recursiveValid(node.left, left, node.val) &&
        this.recursiveValid(node.right, node.val, right);
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;
